### PR TITLE
[Service Bus] .NET Client Library Adoption

### PR DESCRIPTION
**Pull request by jsquire**
## Summary

The goal of this set of changes is to adopt the existing Service Bus client library for .NET from its [current repository](https://github.com/Azure/azure-service-bus-dotnet)  (keeping history) with a focus on ensuring a light touch.  Changes have intentionally been kept to a minimum, with only those needed for compatibility with the local engineering system for builds and testing.  The code and other artifacts otherwise have been left as-is, including the existing code formatting and conventions.

### Repository Structure
  - The [Service Bus area](https://github.com/Azure/azure-sdk-for-net/tree/master/src/SDKs/ServiceBus) has been restructured to separate the management and client areas.

### Service Bus Client 
  - Imported from its previous repository into a dedicated area in this repository following the `data-plane` convention.  

- Projects were adjusted to adapt to the local engineering system to allow for clean builds.  

- Artifacts no longer needed have been cleaned up and removed.  

- Errors from the import and rebase have been fixed, with some minor adjustments made.  

- Infrastructure was added to the tests to allow categorization of `Live` tests, following convention.  

- Test expectations were adjusted to support non-Windows platforms.  

- Tests that are bound to Windows have been marked as `skip` on other platforms.  

### Engineering System
  - An additional parameter was added to the [data plane build properties](https://github.com/Azure/azure-sdk-for-net/blob/master/eng/Directory.Build.Data.props) in order allow test projects to opt-in for signing/strong naming.  This was needed to support the use of the `InternalsVisibleTo` attribute for exposing non-public members to tests.

  - A key intended for signing of test assemblies was generated to allow test libraries to be signed with a dedicated key not associated with the official shared key used for client assemblies.  The test key contains both the public and private keys and is intended for immediate (not delayed) signing.

  - Additional logic was added to the [data plane build properties](https://github.com/Azure/azure-sdk-for-net/blob/master/eng/Directory.Build.Data.props) in order to set appropriate signing properties for test projects to use the new test key.

## Last Upstream Rebase

Tuesday, March 12, 2019  04:11pm (EDT)

## Related and Follow-Up Issues
- [Service Bus Client: Enhance Readme](https://github.com/Azure/azure-sdk-for-net/issues/5349)  (#5349)
- [Service Bus Client: Review and Clean Local Assets](https://github.com/Azure/azure-sdk-for-net/issues/5354) (#5354)

### CODE REVIEW

File: src/SDKs/ServiceBus/data-plane/README.md
Comment (KrzysztofCwalina) [inline src/SDKs/ServiceBus/data-plane/README.md]: This does not seem to follow our REDME format. In particular, we wanted our readmes to start with how to use the package and then much later on talk about how to contribute to the package.

Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/README.md]: I was requested to keep as much as-is as possible for the initial adoption.  This was the ReadMe as it exists in the source repository.  

@AlexGhiondea - Thoughts?  Should we rewrite to snap to local norms or keep with the minimal change approach?

File: README.md
Comment (weshaggard) [inline README.md]: This was added by a recent merge please ensure you merge correctly with it.

File: eng/Directory.Build.Data.targets
Comment (weshaggard) [inline eng/Directory.Build.Data.targets]: Can you help me understand why a test assembly needs to be signed?

Comment (weshaggard) [reply eng/Directory.Build.Data.targets]: If we do have a case for this we should probably use a different key from the MSSharedKey for the signing and ensure we don't get them signed by the signing service officially but only local sn signing.

File: SDKs/ServiceBus/data-plane/Directory.Build.props
Comment (weshaggard) [inline SDKs/ServiceBus/data-plane/Directory.Build.props]: I recently set LangVersion=latest in common props so unless there is a strong need to pin the version I'd exclude this. As for TreatWarningsAsErrors I'm fine with it being here for now but I plan to turn it on globally once I can get all the warnings fixed.

File: src/SDKs/ServiceBus/data-plane/README.md
Comment (weshaggard) [reply src/SDKs/ServiceBus/data-plane/README.md]: This readme does feel more like a how to build/contribute to this library as opposed to how to consume this library also a lot of it applies to more then just this library so I'm with @KrzysztofCwalina that this should be updated but I don't have concerns if it happens as a secondary step.

**Comment by weshaggard** [comment]
Since we initially talked the plan for the overall repo structure change has been approved you can see what the change will look like at https://github.com/Azure/azure-sdk-for-net/issues/5213. This is just an FYI as we will likely do a large pass to update all them at once but I wanted to give you a heads up that the structure will be changing.

File: README.md
Comment (jsquire) [reply README.md]: Rebase to upstream this morning seems to have cleared this.   My goal is to have the opportunity to do a final rebase before the PR gets merged to keep any of this kind of drift to a minimum.

File: eng/Directory.Build.Data.targets
Comment (jsquire) [reply eng/Directory.Build.Data.targets]: In a nutshell, I'm unable to get a clean build without doing so due to quirks in referencing  between signed and unsigned assemblies when a `friend` relationship is applied.  I wish that it weren't so, but am not aware of a way to work around it otherwise.  More detail can be found below.

If you've got a trick to loosen the reference restrictions that I overlooked or an idea for an alternate approach, I'd welcome it so that I can avoid the need to sign.   Likewise, I'd be happy to go with an alternate key if that ends up being the recommendation.   In that case, I'd ask if we would be using a common one at the engineering level or going with ones generated per-SDK where needed.

#### Scenario/Justification
- The client library makes uses of the `InternalsVisibleTo` attribute to expose non-public members to the test suite.  

- The reference to the signed client library is rejected at build time for the unsigned test library, which appears to be compliments of the `friend` designation.  

- Building with both libraries signed or both libraries unsigned works just fine.  Signing the client library and not the test suite results in a build failure.  I did not attempt to sign the test suite and link to an unsigned client library.

- The surface area of the use of internals in the existing test suite is large enough that working around it by either removing the dependency or shifting to a reflection-based approach did not seem justified, given the level of work entailed and potential risk of regressions.

- The goal of these changes is to adopt the client library into this repository with as minimal a set of changes as possible; digging in with additional analysis to validate marking the internal members `public` and understanding the surface area change to the client library did not seem to have justifiable ROI to expand the assigned task.

- While I'd agree with the statement that tests should be focused on the public contract, pragmatically, I'd also agree that it is sometimes helpful or necessary to test against members that you don't want to surface.   Use of the `InternalsVisibleTo` approach is, in my experience, reasonable and a fairly common way of doing so.   Based on that, I can see a need to support it going forward.  In my opinion, doing so via a common approach that allows hinting to the engineering system that signing is needed was a better approach of isolating local overrides on a per-SDK basis.

File: SDKs/ServiceBus/data-plane/Directory.Build.props
Comment (jsquire) [reply SDKs/ServiceBus/data-plane/Directory.Build.props]: These were carried over from the original project file.  I hadn't thought to revisit, but in doing so, the build and test run comes back clean without the version pinned;  taking it out.

File: eng/Directory.Build.Data.targets
Comment (weshaggard) [reply eng/Directory.Build.Data.targets]: For some reason I don't see my follow-up comment, not sure if github messed up or if I forget to hit post, at any rate for completeness I'm pointing out my thoughts on InternalsVisibleTo by linking it here again https://github.com/dotnet/corefx/issues/160. Although based on some of your comments it seems like you saw my earlier comment, anyways.

My dislike of InternalsVisibleTo aside I agree we should have a common mechanism for allowing tests to be signed so I suggest we generate a new test key that contains the full public/private key and commit it to the repo and use that for any test assemblies that we want to strong name sign. I'm happy to add that support or if you like you can take a shot at adding support for it. I'm also happy to do it as a follow-up to this PR and not block on it but I definitely don't want us to start signing test libraries with our MSSharedKey as a long term solution.

File: src/SDKs/ServiceBus/data-plane/README.md
Comment (AlexGhiondea) [reply src/SDKs/ServiceBus/data-plane/README.md]: @jsquire let's file an issue to update the readme after we merge this.

File: eng/Directory.Build.Data.targets
Comment (jsquire) [reply eng/Directory.Build.Data.targets]: I hadn't seen your earlier comment, but it does mirror my own thoughts.   :)

Per our offline discussion, I'll shift to a test-specific full key and adjust the other signing properties to accommodate.

File: src/SDKs/ServiceBus/data-plane/README.md
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/README.md]: Will do.  Taking a note to add it and resolving.

File: eng/Directory.Build.Data.targets
Comment (jsquire) [reply eng/Directory.Build.Data.targets]: Changes have been made.  Signing for test assemblies is now using a newly generated key and setting properties for immediate signing.   Going to resolve, feel free to reopen if there is more discussion needed.

File: src/SDKs/ServiceBus/data-plane/README.md
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/README.md]: Created:  (Service Bus Client: Enhance ReadMe #5349)[https://github.com/Azure/azure-sdk-for-net/issues/5349]

File: eng/Directory.Build.Data.targets
Comment (weshaggard) [inline eng/Directory.Build.Data.targets]: For this special case I think I would delete the line above where you are using `AllowTestAssemblySigning`  and continue to always set SignAssembly=false by default and then instead use a new variable called `SignTestAssembly` and use that in the condition here.

File: eng/Directory.Build.Data.targets
Comment (weshaggard) [inline eng/Directory.Build.Data.targets]: With my suggestions below then you should be able to eliminate the condition changes here.

File: eng/Directory.Build.Data.targets
Comment (weshaggard) [inline eng/Directory.Build.Data.targets]: Looks like a extra space was added to the opening tag I'm surprised that even parses as valid xml.

File: eng/Directory.Build.Data.targets
Comment (jsquire) [reply eng/Directory.Build.Data.targets]: Oooh, I like that.  Much cleaner semantics.   Changes incoming.

File: eng/Directory.Build.Data.targets
Comment (jsquire) [reply eng/Directory.Build.Data.targets]: Thanks!  Good catch.

File: eng/Directory.Build.Data.targets
Comment (weshaggard) [inline eng/Directory.Build.Data.targets]: Yes this is what I was wanting. Not particularly important but I don't think we need the IsTestProject property in the condition.

Comment (jsquire) [reply eng/Directory.Build.Data.targets]: I believe that we do if the properties are being set at the level of the solution's Directory.Build.props and not in the individual project file(s) - which is my current implementation.  I'd like to keep that, as it feels like it should work from either level.

 I experimented with pulling these into the existing `IsTestProject` conditional above, but had to tag each with the `SignTestAssembly` condition due to the inability to nest `PropertyGroup` elements.  I also experimented with another block for "if this isn't a test assembly force the `SignTestAssembly` to false.

I'll admit that I'm not thrilled with the compound condition, but it seemed the lesser of evils compared against the other experiments.  At least with this approach the intent is clear at a glance instead of having some implicit resets floating in the environment that are potentially non-obvious.

Comment (weshaggard) [reply eng/Directory.Build.Data.targets]: The case I had in mind was the IsTestLibraryHelper that might need to be signed but we can handle that if and when the case ever occurs. For now I'm fine with what you have.

**Review by weshaggard — APPROVED** [review]
Initial port looks good but there are definitely some things to clean-up afterwards. Things like the extra .gitignore file and such which isn't needed, but I'm fine with those clean-up items being done later.

File: eng/Directory.Build.Data.targets
Comment (jsquire) [reply eng/Directory.Build.Data.targets]: I missed that case.   Given that, I think it may be better to go with forced unset of not a test or helper project.   I'm going to throw that in.

**Comment by jsquire** [comment]
@weshaggard 
> Initial port looks good but there are definitely some things to clean-up afterwards. Things like the extra .gitignore file and such which isn't needed, but I'm fine with those clean-up items being done later.

Just a note that I intentionally left the local `.gitignore` in place and cleaned out redundant items.  There were some Service Bus specifics that I wasn't convinced should be polluting the root ignore file, such as:

```
test/Microsoft.Azure.ServiceBus.UnitTests/API/ApiApprovals.ApproveAzureServiceBus.received.txt
```

Ideally, in this case, the test suite should clean up its own mess but I can see other cases where it may be worth a discussion to establish a consistent practice around how to handle ignoring of things that are granular to a specific solution.

**Review by weshaggard — APPROVED** [review]

**Comment by weshaggard** [comment]
>Just a note that I intentionally left the local .gitignore in place and cleaned out redundant entries from it. There were some Service Bus specifics that I wasn't convinced should be polluting the root ignore file, such as:

I'm not much of a fan having these .gitignore files throughout the source tree as it leads to each sdk having something different and inconsistent. We should try to make the file types we ignore consistent across the repo and then put that pattern in the root gitignore file.

**Comment by jsquire** [comment]
> I'm not much of a fan having these .gitignore files throughout the source tree as it leads to each sdk having something different and inconsistent. We should try to make the file types we ignore consistent across the repo and then put that pattern in the root gitignore file.

Works for me.   Added [Service Bus Client: Review and Clean Local Assets #5354](https://github.com/Azure/azure-sdk-for-net/issues/5354) to track this.

**Pull request merged**
